### PR TITLE
Add String.prototype.padStart and String.prototype.padEnd

### PR DESCRIPTION
**Pull request by regiskuckaertz**
Hey, those ones seem very useful, I figured I could give it a try.

### CODE REVIEW

File: polyfills/String/prototype/padStart/tests.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+it('has correct instance', function () {
+	proclaim.isInstanceOf(String.prototype.padStart, Function);
+});
+
+it('has correct argument length', function () {
+	proclaim.equal(String.prototype.padStart.length, 2);
+});
+
+it('works with strings', function () {
+	proclaim.equal('a'.padStart(0, 'a'), 'a');
```
Comment (JakeChampion) [inline polyfills/String/prototype/padStart/tests.js]: May we also test negative lengths and undefined lengths please. They should both be no-ops.

File: polyfills/String/prototype/padStart/tests.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+	proclaim.equal('a'.padStart(1, 'a'), 'a');
+	proclaim.equal('a'.padStart(5, 'a'), 'aaaaa');
+	proclaim.equal('a'.padStart(5, 'bc'), 'bcbca');
+	proclaim.equal('a'.padStart(5, 'bcdef'), 'bcdea');
```
Comment (JakeChampion) [inline polyfills/String/prototype/padStart/tests.js:18]: May we also test non-string `padString` arguments. I believe they should get stringified.

Comment (JakeChampion) [reply polyfills/String/prototype/padStart/tests.js:18]: We also need to test calls which do not pass in a `padString`, `padString` should default to a space.

**Comment by JakeChampion** [comment]
Looks good @regiskuckaertz, thanks for doing this! :-)

**Comment by regiskuckaertz** [comment]
Hi @JakeChampion Thanks for your review, I completely forgot about these cases 😊

File: polyfills/String/prototype/padEnd/config.json [hunk 3]
```diff
@@ -0,0 +1,18 @@
+{
+	"aliases": [
+		"es7"
+	],
+	"browsers": {
+		"android": "*",
+		"bb": "*",
+		"chrome": "<57",
+		"firefox": "<51",
```
Comment (JakeChampion) [inline polyfills/String/prototype/padEnd/config.json]: According to MDN, Firefox has had padEnd since version 48 -- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/padEnd#Browser_compatibility

File: polyfills/String/prototype/padEnd/config.json [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+		"opera": "<44",
+		"safari": "<10",
+		"ie": "<15",
+		"ie_mob": "*",
+		"ios_saf": "<10",
+		"firefox_mob": "<51"
```
Comment (JakeChampion) [inline polyfills/String/prototype/padEnd/config.json]: Same here, Firefox mobile apparently has had padEnd since version 48

**Comment by regiskuckaertz** [comment]
Great catch! I took the version out of MDN too, but they got it wrong in [padStart](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/padStart)---have submitted a patch there too.

**Comment by JakeChampion** [comment]
@regiskuckaertz The polyfill and/or tests are failing in some browsers:

```
- chrome/56:
    -> String.prototype.padEnd
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=String.prototype.padEnd
    -> String.prototype.padStart
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=String.prototype.padStart
 - chrome/48:
    -> String.prototype.padEnd
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=String.prototype.padEnd
    -> String.prototype.padStart
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=String.prototype.padStart
 - firefox/52:
    -> String.prototype.padEnd
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=String.prototype.padEnd
    -> String.prototype.padStart
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=String.prototype.padStart
 • firefox/44 (targeted): Error: Timed out at 'polling for results'
 - ie/14:
    -> String.prototype.padEnd
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=String.prototype.padEnd
    -> String.prototype.padStart
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=String.prototype.padStart
 - ie/13:
    -> String.prototype.padEnd
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=String.prototype.padEnd
    -> String.prototype.padStart
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=String.prototype.padStart
 - ie/11:
    -> String.prototype.padEnd
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=String.prototype.padEnd
    -> String.prototype.padStart
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=String.prototype.padStart
 - ie/10:
    -> String.prototype.padEnd
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=String.prototype.padEnd
    -> String.prototype.padStart
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=String.prototype.padStart
 - ie/9:
    -> String.prototype.padEnd
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=String.prototype.padEnd
    -> String.prototype.padStart
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=String.prototype.padStart
 - ie/8:
    -> String.prototype.padEnd
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=String.prototype.padEnd
    -> String.prototype.padStart
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=String.prototype.padStart
 - ie/7:
    -> String.prototype.padEnd
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=String.prototype.padEnd
    -> String.prototype.padStart
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=String.prototype.padStart
 - safari/9:
    -> String.prototype.padEnd
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=String.prototype.padEnd
    -> String.prototype.padStart
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=String.prototype.padStart
 - safari/8:
    -> String.prototype.padEnd
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=String.prototype.padEnd
    -> String.prototype.padStart
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=String.prototype.padStart
 - android/4.4:
    -> String.prototype.padEnd
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=String.prototype.padEnd
    -> String.prototype.padStart
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=String.prototype.padStart
```

**Comment by regiskuckaertz** [comment]
Hi @JakeChampion 

Sorry about that, was in a hurry last Friday and missed a typo in one of my tests. I ran them across several browsers, everything looks good now.

File: polyfills/String/prototype/padEnd/config.json [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+		"es7"
+	],
+	"browsers": {
+		"android": "*",
+		"bb": "*",
+		"chrome": "<57",
+		"firefox": "<48",
+		"opera": "<44",
+		"safari": "<10",
+		"ie": "<15",
```
Comment (leggsimon) [inline polyfills/String/prototype/padEnd/config.json:12]: @JakeChampion I'm assuming this just includes Edge in with IE rather that having them as "separate" browsers?

File: polyfills/String/prototype/padEnd/polyfill.js [hunk 6]
```diff
@@ -0,0 +1,17 @@
+(function () {
+	Object.defineProperty(String.prototype, 'padEnd', {
+		configurable: true,
+		enumerable: false,
```
Comment (leggsimon) [inline polyfills/String/prototype/padEnd/polyfill.js:4]: @JakeChampion This is so that it doesn't then somehow get polyfilled twice? I don't understand why the property should be marked as non-enumerable? What doesn't that tell us?

**Review by onishiweb — APPROVED** [review]

File: polyfills/String/prototype/padEnd/polyfill.js [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+		value: function padEnd (targetLength, padString) {
```
Comment (leggsimon) [inline polyfills/String/prototype/padEnd/polyfill.js:5]: This may be a silly question/observation but I thought the `target` was normally the thing the method was being called on?
But actually as I write this comment out I realise that it actually makes sense that the target whould be the thing you are trying to achieve 🤦‍♂️ (I'm going to keep this comment as written proof that I do realise things)

File: polyfills/String/prototype/padEnd/config.json (hunk 5 above)
Comment (JakeChampion) [reply polyfills/String/prototype/padEnd/config.json:12]: Correct :-)

File: polyfills/String/prototype/padEnd/polyfill.js (hunk 6 above)
Comment (JakeChampion) [reply polyfills/String/prototype/padEnd/polyfill.js:4]: It stops the method from appearing in the results of any enumeration such as `for..in`.

```js
➜  ~ node
> for (var a in String.prototype) {
...   console.log(a);
... }
undefined
>
> String.prototype.padEnd = function padEnd (targetLength, padString) {
...   targetLength = targetLength | 0;
...   if (targetLength <= this.length) return String(this);
...   padString = String(padString || " ");
...   var repeat = Math.ceil((targetLength - this.length) / padString.length);
...   while (repeat--) {
.....     padString += padString;
.....   }
...   return String(this) + padString.substr(0, targetLength - this.length);
... };
[Function: padEnd]
>
> for (var b in String.prototype) {
...   console.log(b);
... }
padEnd
undefined
>
> delete String.prototype.padEnd;
true
>
> Object.defineProperty(String.prototype, 'padEnd', {
...   configurable: true,
...   enumerable: false,
...   value: function padEnd (targetLength, padString) {
.....     targetLength = targetLength | 0;
.....     if (targetLength <= this.length) return String(this);
.....     padString = String(padString || " ");
.....     var repeat = Math.ceil((targetLength - this.length) / padString.length);
.....     while (repeat--) {
.......       padString += padString;
.......     }
.....     return String(this) + padString.substr(0, targetLength - this.length);
.....   },
...   writable: true
... });
[String: '']
>
> for (var c in String.prototype) {
...   console.log(c);
... }
undefined
>
```

Comment (leggsimon) [reply polyfills/String/prototype/padEnd/polyfill.js:4]: Ahhhhhhhh. Ok I see 👍

File: polyfills/String/prototype/padEnd/polyfill.js [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+			targetLength = targetLength | 0;
```
Comment (rowanmanning) [inline polyfills/String/prototype/padEnd/polyfill.js:6]: Using bitwise OR here to floor `targetLength` could cause issues with particularly large numbers. E.g.

```js
3000000000 | 0 === -1294967296
```

In the case above the resulting string reaches the maximum length in Chrome Canary and Firefox stable before this becomes an issue. However as [maximum string length can differ between different JavaScript engines](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Errors/Resulting_string_too_large) this could feasibly cause problems in the future.

Not sure whether it's a pressing issue but worth documenting. An alternative could be to use:

```js
targetLength = Math.floor(parseInt(targetLength, 10)) || 0;
```

**Review by rowanmanning — APPROVED** [review]

**Review by JakeChampion — APPROVED** [review]

**Comment by JakeChampion** [comment]
Thanks for the PR! 

Did you know we offer one month's free premium access to all FT content (normally $50) to anyone who has a pull request merged in this repository? If you want to claim it, there's [one Google form](https://docs.google.com/a/ft.com/forms/d/1aykPiYBwqKpKY6nvGVi43wOU_5epr3XLgE00Vi52SMQ/viewform) to fill out :-)

**Comment by regiskuckaertz** [comment]
Cheers 😃  I didn't know, that's fantastic! Thanks for the tip, I'll fill it in right away.
